### PR TITLE
enable parallel executor if inter_op_num_threads is set

### PR DESCRIPTION
**Pull request by yufenglee**
**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/test/perftest/command_args_parser.cc [hunk 1]
```diff
@@ -127,6 +127,8 @@ namespace perftest {
         test_config.run_config.inter_op_num_threads = static_cast<int>(OrtStrtol<PATH_CHAR_TYPE>(optarg, nullptr));
         if (test_config.run_config.inter_op_num_threads < 0) {
           return false;
+        } else {
+          test_config.run_config.enable_sequential_execution = false;
```
Comment (KeDengMS) [inline onnxruntime/test/perftest/command_args_parser.cc]: Should parallel executor only get enabled when inter_op_num_threads > 1?

**Comment by pranavsharma** [comment]
Currently the way to enable parallel executor is to disable sequential executor by calling the DisableSequentialExecution API. I believe this is not intuitive at all and something we should fix before 1.0 release. The code today honors the inter_op_num_threads option only if DisableSequentialExecution API has been called. I think what we should do is 
- Eliminate the EnableSequentialExecution/DisableSequentialExecution APIs
- Expose 2 new APIs EnableParallelExecution/DisableParallelExecution
- Honor the inter_op_num_threads option only if the above API has been called on the session.

File: onnxruntime/test/perftest/command_args_parser.cc (hunk 1 above)
Comment (yufenglee) [reply onnxruntime/test/perftest/command_args_parser.cc]: If inter_op_num_threads=0, it uses the default std::max<int>(1, std::thread::hardware_concurrency() / 2);

so you mean we need to exclude inter_op_num_threads=1?

**Comment by yufenglee** [comment]
> Currently the way to enable parallel executor is to disable sequential executor by calling the DisableSequentialExecution API. I believe this is not intuitive at all and something we should fix before 1.0 release. The code today honors the inter_op_num_threads option only if DisableSequentialExecution API has been called. I think what we should do is
> 
> * Eliminate the EnableSequentialExecution/DisableSequentialExecution APIs
> * Expose 2 new APIs EnableParallelExecution/DisableParallelExecution
> * Honor the inter_op_num_threads option only if the above API has been called on the session.

Agree on the 1 and 2. For 3, shouldn't we honor inter_op_num_threads if EnableParallelExecution/DisableParallelExecution not called? If i set inter_op_num_threads >1 explicitly, i mean to use parallel executor.

**Comment by KeDengMS** [comment]
Please also consider extension in the future to mode more than Sequential/Parallel. To me the inter_op_num_threads just means the number of threads to use for model parallelism, not necessary for just parallel executor. We may have something like PipelineExecutor for example.

---
In reply to: [538583289](https://github.com/microsoft/onnxruntime/pull/2000#issuecomment-538583289) [](ancestors = 538583289)

File: onnxruntime/test/perftest/command_args_parser.cc (hunk 1 above)
Comment (KeDengMS) [reply onnxruntime/test/perftest/command_args_parser.cc]: The default value of 0 is kinda ambiguous with this change then. I think we meant to default to sequential executor, so maybe we should change inter_op_num_threads default value to 1?

---
In reply to: [331709085](https://github.com/microsoft/onnxruntime/pull/2000#discussion_r331709085) [](ancestors = 331709085)

**Comment by snnn** [comment]
Any updates?

**Comment by yuslepukhin** [comment]
I believe enabling parallel execution must only be done explicitly by the user. The reason is we do not use memory planning and some even simple models run out of memory in parallel execution. I once spent debugging several days a crash only to find out that it was running out of memory.

**Pull request closed without merging**